### DISCUSSION TITLE
The future of Logbook

### DISCUSSION THREAD

**Discussion by whiskeysierra** [zalando/logbook#1379] [General]
This discussion is an attempt to consolidate questions and suggestions made in different, currently open PRs about the status and future of this project.

/cc @agebhar1 @phejl

**Comment by whiskeysierra** [comment]
As said in https://github.com/zalando/logbook/pull/1375#issuecomment-1324878342, I'd be willing to help maintaining a fork.
That fork could possibly be something that people will perceive as an official successor project.

I saw that the "logbook" org is already taken, but I reached out to see if we could get it:
https://github.com/logbook/logbook.github.io/issues/1

**Reply by phejl** [reply]
I'm ready to help. What is the plan? Wait a bit if we get a response from "logbook" org?

**Reply by whiskeysierra** [reply]
Yeah, but nothing yet...

Any other suggestions about org name?
I mean I could also host a fork over at github.com/whiskeysierra, but that would raise the immediate next question: What do we pick as a groupId for maven central (it has to be a domain that you own, e.g. a github page).

**Reply by whiskeysierra** [reply]
See https://central.sonatype.org/publish/#individual-projects-open-source-software-repository-hosting-ossrh

> you must choose a groupId for a Web Domain that you own

**Reply by agebhar1** [reply]
Hi, Github is also fine

> In the case of Github, Gitee, Bitbucket or similar groupIds (io.github.username), you must create a temporary, public repository named OSSRH-TICKETNUMBER. Check the details in the [Coordinates document](https://central.sonatype.org/publish/requirements/coordinates/#supported-code-hosting-services-for-personal-groupid).
>
> github.com/yourusername -> io.github.yourusername

Did that once with https://github.com/agebhar1/micrometer-certs-expiration.

**Reply by whiskeysierra** [reply]
Yeah, it is. It just boils down to the same challenge then: Finding a good place to host the code 😉

**Reply by agebhar1** [reply]
I suggest you @whiskeysierra, should host the code, maybe with @phejl and me as collaborators. Because you already know and maintained the project.

**Reply by whiskeysierra** [reply]
@otrosien Given that Zalando no longer maintains this project, can we transfer the ownership (instead of creating a fork)?

**Comment by whiskeysierra** [comment]
@agebhar1 Let's create a separate thread in here about maven coordinates.

Let's say we settle on my github space for logbook, i.e. whiskeysierra/logbook.
That kind of implies we also change the groupId from org.zalando.logbook to io.github.whiskeysierra.logbook. Right?

The top-level package is a different story, where I'd actually suggest to use just `logbook` (no reverse DNS).

**Reply by agebhar1** [reply]
whiskeysierra/logbook ✔️ 
io.github.whiskeysierra.logbook ✔️

**Comment by marcindabrowski** [comment]
Why not create a new organization, for example, `logbook-java` or something similar?

**Reply by whiskeysierra** [reply]
I reserved logbook-http for now. And we should get back from Zalando about whether we get the repo transferred or whether we need to fork it.

**Reply by agebhar1** [reply]
@whiskeysierra for how long should we wait for a response?

**Reply by whiskeysierra** [reply]
I got 14th of December as when their Opensource group meets next.

**Reply by agebhar1** [reply]
It's the 18th of December. How should we proceed?

**Reply by leviferreira** [reply]
Hi everyone, I'm Levi from Zalando. I'm sorry that the project might feel a bit abandoned. The Zalando engineering community have been working hard to revive some of our open source projects. Answering your questions we've decided to still maintain the logbook open-source project for now and are working to setup a internal group of maintainers that will keep the project in a healthy state, with a better support for the open source community like we have in other projects. This way the open source community can continue to contribute to the project. Of course you are free to fork the project as the project is under MIT license, in case you decide to proceed with your fork.

**Reply by hemju** [reply]
Hi Levi, thanks for the update. Could you elaborate on what `still maintain` actually means? Will you and your team be actively working on upgrades/new features? There are 27 Pull Requests open and need attention.

There are other questions as well; currently, the project is not compatible with Spring Boot 3, which is, in my opinion, a big problem (especially with an open PR to fix that). So, for example, will Zalando merge this PR to support Spring Boot 3 even if you internally use Spring Boot 2?

Can you please share a timeline for when this new support will arrive and when updates will come in?

**Reply by leviferreira** [reply]
@hemju 
> Will you and your team be actively working on upgrades/new features? There are 27 Pull Requests open and need attention.

Yes, as soon as we have set the internal group of maintainers.

> So, for example, will Zalando merge this PR to support Spring Boot 3 even if you internally use Spring Boot 2?

Yes, as soon as we have set the internal group of maintainers.

> Can you please share a timeline for when this new support will arrive and when updates will come in?

I expect the support to arrive in January '23.

**Reply by hemju** [reply]
January 2023 sounds great. Thank for the fast reply.

**Reply by sik00r** [reply]
Hi, that's already one third od January 2023, what is the progress with Spring boot 3?

**Reply by SledgeHammer01** [reply]
> Hi, that's already one third od January 2023, what is the progress with Spring boot 3?

@whiskeysierra I'd have to say, if we haven't heard anything by now just fork it and move on with our lives. Same thing happened with the highly used Spring Fox project where it was abandoned and the author "promised" things and while he did deliver on his initial "promise", it was abandoned again soon after,

**Reply by hemju** [reply]
I agree with @SledgeHammer01. Considering current communication, I would have more trust in a fork than being dependent on the company policy of Zalando. When this project is no longer important for Zalando (as it looks like), then the maintenance will be minimal at best. I would support a fork (the name `logbook-http` sounds nice) and maybe I can help with the transition as well.

**Reply by whiskeysierra** [reply]
I'm inclined to agree, but I was really hoping that we could transfer the ownership of the repository instead of creating an independent fork. Not only would that move all the historic PRs/issues but also the currently open ones. And it would redirect people to the official successor project. Without this, we'd basically start off at zero in terms of stars/downloads/awareness/etc.

**Reply by whiskeysierra** [reply]
In addition, if Zalando (/cc @leviferreira) would agree to an ownership transfer, we could also discuss putting in a maven relocation pom to guide maven users over to the new one: https://maven.apache.org/guides/mini/guide-relocation.html

**Reply by hemju** [reply]
ping @leviferreira any thoughts on the ownership transfer

**Reply by leviferreira** [reply]
Hi everyone, We are still committed as a company to maintain logbook and are not planning any official transfer of ownership. The new maintainers are being prepared and we should should be able to let you know who they are by end of next week. I'll keep you updated when I have more info.

**Reply by leviferreira** [reply]
Hi Everyone, 

The people to maintain logbook is now selected and they are getting some onboarding and access next week. 
There will be 6 colleagues maintaining and contributing so I expect already some action next week. 
The initial goal is to make sure the PR's open are getting attention.

**Reply by agebhar1** [reply]
Hi @leviferreira, thank you for the update.

**Reply by hemju** [reply]
Really cool. Thank you for the update.

**Reply by jklingstedtt** [reply]
"Next week" was almost two weeks ago. Any progress on this?

**Reply by leviferreira** [reply]
Hey @jklingst The team is already working on it since two weeks now. You can see movements in PRs, etc.

**Reply by foobar-1337** [reply]
@leviferreira Do you have an estimated release date for these fixes?

**Reply by nhmarujo** [reply]
Hi @leviferreira 👋 I can see the project is getting some love again and is good to see it, so thank you all for that. I see version `2.14.0` was released LONG ago (`Oct 12, 2021`) and many things are merged on `master` that are not yet part of any release. This has been like that even before you started maintaining the project now.
I understand everyone here is craving for the SB3 support (ofc I am too) and you guys are kinda tidying up the house and paving the path to get there, so it takes a bit. But in the meantime, is there any planned release so we can get all the goodies that were already merged so far? Thanks

**Reply by kasmarian** [reply]
Hi @nhmarujo, yes, we do plan to do the 2.15.0 soon. To let all the clients get the latest updates before making any big steps, like SB3 support.

**Reply by nhmarujo** [reply]
Hi @kasmarian , thank you. That makes perfect sense. Looking forward to it 😄

**Comment by whiskeysierra** [comment]
If you come up with a good name, I'm all for it. Java is to restrictive,
I'd say because with modern Spring there is quite some Kotlin projects out
there that can an do still use Logbook.

What about logbook-jvm?

**Reply by marcindabrowski** [reply]
You are right. I forgot that java is a trademark of Oracle and that could be a problem.
What about http-logbook, or logbook-http?

**Reply by agebhar1** [reply]
:laughing: I had exactly the same idea.

**Reply by phejl** [reply]
`logbook-http` sounds great to me.

**Comment by whiskeysierra** [comment]
Yeah, agreed. It's so bound to HTTP, I don't ever see it expanding beyond
that protocol.

**Comment by maciejwalkowiak** [comment]
@whiskeysierra I am just checking if anything moved forward..? In case you need some support with Spring modules let me know.

**Reply by Ahli** [reply]
When you scroll up a bit and check other comments, you will see [this](https://github.com/zalando/logbook/discussions/1379#discussioncomment-4797824) posted last Friday:

> Hi Everyone,
> 
> The people to maintain logbook is now selected and they are getting some onboarding and access next week. There will be 6 colleagues maintaining and contributing so I expect already some action next week. The initial goal is to make sure the PR's open are getting attention.

**Comment by hemju** [comment]
According to this comment: https://github.com/zalando/logbook/pull/1417#issuecomment-1473715463 there is no ETA when logbook will support Spring 3. Going back to my comment here: https://github.com/zalando/logbook/discussions/1379#discussioncomment-4703842

As a community we depend on how Zalando actually uses Logbook internally, and not on what the community vision/requirements are. E.g. my guess is Zalando will have a Spring Boot 3 support when they upgrade to Spring Boot 3 internally. Not earlier. I do believe that something like Logbook would be a useful project for a wide variety of companies. Especially, when the requirements are not dictated by one consumer. So I want to raise again the discussion of created a community fork of Logbook. Is that still a possibility?

@maciejwalkowiak @whiskeysierra @agebhar1 @marcindabrowski & others

Btw. just to be clear, I really appreciate the work of all collaborators and also from Zalando to make Logbook Open Source in the first place.

**Reply by maciejwalkowiak** [reply]
I understand your concerns. Forking is easy, maintaining the fork is a challenge. In terms of community work i can help by providing an advice if such is needed, but for anything bigger would mean a paid contract the me. Feel free to drop me an email or message on Twitter if your company would be interested.

**Reply by agebhar1** [reply]
Hi,

I've updated the PR https://github.com/zalando/logbook/pull/1375 to the current master. As already mentioned in https://github.com/zalando/logbook/discussions/1426 "The way to move forward with Spring 6 / Spring Boot 3" it's not clever and maintainable to copy ~5k of extra code. I hope to find some time next weekend to implement @maciejwalkowiak idea https://github.com/zalando/logbook/discussions/1426#discussioncomment-5069937 to use Maven Shade.

**Reply by agebhar1** [reply]
Already done in #1467 by @ChristianLohmann.

**Comment by ChristianLohmann** [comment]
Regarding Logbooks Spring 6 support there is a new (actually reopened) pull request #1467 as @agebhar1 already mentioned which enables logbook finally to run with Spring 6. Thanks to @maciejwalkowiak additional Spring 5 backwards compatibility has been added using a shaded `logbook-servlet` artifact. That reduced the effort to, more or less only one additional AutoConfiguration class.    
We are looking forward to get this done very soon.

**Reply by ChristianLohmann** [reply]
Short Update: #1467 has been merged. The next version which is expected to be released within the next days will be the one then.

**Reply by hemju** [reply]
Thank you!

**Reply by nhmarujo** [reply]
Hi. Is the release of 3.0.0 happening any time soon? 😄 Thanks

**Reply by ChristianLohmann** [reply]
Hi, we already released verson 3 as a [release candidate](https://github.com/zalando/logbook/releases/tag/3.0.0-RC.0)

**Reply by nhmarujo** [reply]
Hi @ChristianLohmann . Yes, I noticed that and we have already used it on our own RCs. But are you suggesting to use your RC version on a production environment?

**Reply by ChristianLohmann** [reply]
For the time being if you already use Spring 6 in production envs you can use that release. We  are planning to add some more breaking changes/features into the next RC which have been a bit forgotten.

**Reply by nhmarujo** [reply]
We are not using Spring 6 in production yet. We have everything ready in RC branches but we have been waiting for this to move to SB3. Can you please clarify what those breaking changes/features are about? Thanks

**Reply by ChristianLohmann** [reply]
We are planning to introduce JDK9 module support and Ktor 2.x support. If you don't use Spring 6 in prod yet then i suggest you to wait at least for the next RC as there is a fix to enable Spring 5 backwards compatibility in spring-security backed projects

**Reply by nhmarujo** [reply]
Thank for the feedback. Do you have any ETA for the next RC?
